http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SslContextProvider.java

```diff
@@ -37,30 +40,46 @@
 public final class SslContextProvider {
     private static final NettyClientLogger log = NettyClientLogger.getLogger(SslContextProvider.class);
     private final Protocol protocol;
+    private final ProtocolNegotiation protocolNegotiation;
     private final SslProvider sslProvider;
     private final TrustManagerFactory trustManagerFactory;
     private final KeyManagerFactory keyManagerFactory;
 
-    public SslContextProvider(NettyConfiguration configuration, Protocol protocol, SslProvider sslProvider) {
+    public SslContextProvider(NettyConfiguration configuration, Protocol protocol, ProtocolNegotiation protocolNegotiation,
+                              SslProvider sslProvider) {
         this.protocol = protocol;
+        this.protocolNegotiation = protocolNegotiation;
         this.sslProvider = sslProvider;
         this.trustManagerFactory = getTrustManager(configuration);
         this.keyManagerFactory = getKeyManager(configuration);
     }
 
     public SslContext sslContext() {
         try {
-            return SslContextBuilder.forClient()
-                                    .sslProvider(sslProvider)
-                                    .ciphers(getCiphers(), SupportedCipherSuiteFilter.INSTANCE)
-                                    .trustManager(trustManagerFactory)
-                                    .keyManager(keyManagerFactory)
-                                    .build();
+            SslContextBuilder builder = SslContextBuilder.forClient()
+                                                         .sslProvider(sslProvider)
+                                                         .ciphers(getCiphers(), SupportedCipherSuiteFilter.INSTANCE)
+                                                         .trustManager(trustManagerFactory)
+                                                         .keyManager(keyManagerFactory);
+
+            if (protocolNegotiation == ProtocolNegotiation.ALPN) {
+                builder.applicationProtocolConfig(
+                    new ApplicationProtocolConfig(ApplicationProtocolConfig.Protocol.ALPN,
+                                                  ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                                                  ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
```

Comment:
Question: how is `SelectorFailureBehavior` different from `SelectedListenerFailureBehavior`? 
Should we use `FATAL_ALERT`?

Reply:
Seems like `SelectorFailureBehavior` comes first, is the peer in the connection. `SelectedListenerFailureBehavior` is for a configured listener. Updating to `FATAL_ALERT` for both

Reply:
Turns out OpenSsl Provider does not support `FATAL_ALERT`
https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java#L767

Reverting back to original values if OpenSsl provider used, added tests